File: Pletor-canvas/backend/src/services/workflowyApiClient.ts
import { AppError } from '../middlewares/errorHandler.js'

const WORKFLOWY_API = 'https://workflowy.com/api/v1'

// Struktura uzlu z WorkFlowy API
export interface WorkFlowyNode {
  id: string
  name: string
  note: string | null
  priority: number
  data?: { layoutMode?: string }
  createdAt: number
  modifiedAt: number
  completedAt: number | null
}

// Export — ploché pole se vztahy
export interface WorkFlowyExportNode extends WorkFlowyNode {
  parent_id: string | null
}

// Rekonstruovaný strom
export interface WorkFlowyTreeNode extends WorkFlowyNode {
  children: WorkFlowyTreeNode[]
}

// WorkFlowy API klient — oficiální REST API v1
export class WorkFlowyApiClient {
  private apiKey: string
  private cache: { nodes: WorkFlowyExportNode[]; timestamp: number } | null = null
  private cacheTtl = 60_000 // 60 sekund

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const options: RequestInit = {
      method,
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
    }
    if (body) options.body = JSON.stringify(body)

    const response = await fetch(`${WORKFLOWY_API}${path}`, options)

    if (!response.ok) {
      if (response.status === 401 || response.status === 403) {
        throw new AppError(401, 'WORKFLOWY_AUTH_FAILED', 'WorkFlowy API klíč je neplatný')
      }
      if (response.status === 429) {
        throw new AppError(429, 'WORKFLOWY_RATE_LIMIT', 'WorkFlowy API rate limit — zkus to za minutu nebo použij cache')
      }
      throw new AppError(502, 'WORKFLOWY_ERROR', `WorkFlowy API chyba: ${response.status}`)
    }

    return response.json()
  }

  // Načte celý strom — zkusí export, při rate limitu fallback na rekurzivní /nodes
  async getTree(forceRefresh = false): Promise<WorkFlowyTreeNode[]> {
    try {
      const flat = await this.getAllNodesFlat(forceRefresh)
      return buildTree(flat)
    } catch (err) {
      // Při rate limitu fallback na rekurzivní načítání
      if (err instanceof AppError && err.code === 'WORKFLOWY_RATE_LIMIT') {
        return this.getTreeRecursive('None')
      }
      throw err
    }
  }

  // Načte ploché pole všech uzlů přes /nodes-export (rate limit: 1/min, proto cache)
  async getAllNodesFlat(forceRefresh = false): Promise<WorkFlowyExportNode[]> {
    if (!forceRefresh && this.cache && Date.now() - this.cache.timestamp < this.cacheTtl) {
      return this.cache.nodes
    }

    const response = await this.request<{ nodes: WorkFlowyExportNode[] }>('GET', '/nodes-export')
    this.cache = { nodes: response.nodes, timestamp: Date.now() }
    return response.nodes
  }

  // Rekurzivní načítání stromu přes /nodes (bez rate limitu, ale pomalejší)
  private async getTreeRecursive(parentId: string, depth = 0): Promise<WorkFlowyTreeNode[]> {
    if (depth > 5) return [] // Ochrana proti příliš hlubokému zanořování

    const children = await this.getChildren(parentId)
    const result: WorkFlowyTreeNode[] = []

    for (const child of children) {
      const grandchildren = await this.getTreeRecursive(child.id, depth + 1)
      result.push({ ...child, children: grandchildren })
    }

    return result.sort((a, b) => a.priority - b.priority)
  }

  // Načte konkrétní uzel
  async getNode(id: string): Promise<WorkFlowyNode> {
    const response = await this.request<{ node: WorkFlowyNode }>('GET', `/nodes/${id}`)
    return response.node
  }

  // Načte děti uzlu
  async getChildren(parentId: string): Promise<WorkFlowyNode[]> {
    const response = await this.request<{ nodes: WorkFlowyNode[] }>('GET', `/nodes?parent_id=${parentId}`)
    return response.nodes
  }

  // Vytvoří nový uzel
  async createNode(parentId: string, name: string, note?: string): Promise<{ item_id: string }> {
    return this.request<{ item_id: string }>('POST', '/nodes', {
      parent_id: parentId,
      name,
      note,
    })
  }

  // Aktualizuje uzel
  async updateNode(id: string, data: { name?: string; note?: string }): Promise<{ status: string }> {
    return this.request<{ status: string }>('POST', `/nodes/${id}`, data)
  }

  // Smaže uzel
  async deleteNode(id: string): Promise<{ status: string }> {
    return this.request<{ status: string }>('DELETE', `/nodes/${id}`)
  }

  // Invaliduje cache
  invalidateCache() {
    this.cache = null
  }

  // Najde uzel v plochém poli podle ID
  async findNodeById(id: string): Promise<WorkFlowyExportNode | null> {
    const nodes = await this.getAllNodesFlat()
    return nodes.find((n) => n.id === id) ?? null
  }

  // Najde uzel podle textu v názvu
  async findNodeByName(name: string): Promise<WorkFlowyExportNode | null> {
    const nodes = await this.getAllNodesFlat()
    const lower = name.toLowerCase()
    return nodes.find((n) => n.name.toLowerCase().includes(lower)) ?? null
  }
}

// Rekonstruuje strom z plochého pole
function buildTree(nodes: WorkFlowyExportNode[]): WorkFlowyTreeNode[] {
  const map = new Map<string, WorkFlowyTreeNode>()
  const roots: WorkFlowyTreeNode[] = []

  // Vytvoř mapování ID → uzel s prázdnými dětmi
  for (const node of nodes) {
    map.set(node.id, { ...node, children: [] })
  }

  // Propoj rodiče s dětmi
  for (const node of nodes) {
    const treeNode = map.get(node.id)!
    if (node.parent_id && map.has(node.parent_id)) {
      map.get(node.parent_id)!.children.push(treeNode)
    } else {
      roots.push(treeNode)
    }
  }

  // Seřaď děti podle priority
  for (const node of map.values()) {
    node.children.sort((a, b) => a.priority - b.priority)
  }

  return roots.sort((a, b) => a.priority - b.priority)
}

File: Pletor-canvas/backend/app/services/workflowy_api_client.py
import time
from dataclasses import dataclass, field
from typing import Any

import httpx

from app.middlewares.error_handler import AppError

WORKFLOWY_API = "https://workflowy.com/api/v1"


@dataclass
class WorkFlowyNode:
    id: str
    name: str
    note: str | None
    priority: int
    created_at: int
    modified_at: int
    completed_at: int | None = None
    data: dict[str, Any] | None = None


@dataclass
class WorkFlowyExportNode(WorkFlowyNode):
    parent_id: str | None = None


@dataclass
class WorkFlowyTreeNode(WorkFlowyNode):
    children: list["WorkFlowyTreeNode"] = field(default_factory=list)


class WorkFlowyApiClient:
    def __init__(self, api_key: str) -> None:
        self._api_key = api_key
        self._cache: dict[str, Any] | None = None
        self._cache_timestamp: float = 0
        self._cache_ttl = 60.0  # 60 sekund

    async def _request(self, method: str, path: str, body: dict[str, Any] | None = None) -> Any:
        headers = {"Authorization": f"Bearer {self._api_key}", "Content-Type": "application/json"}

        async with httpx.AsyncClient() as client:
            response = await client.request(method, f"{WORKFLOWY_API}{path}", headers=headers, json=body)

        if response.status_code == 401 or response.status_code == 403:
            raise AppError(401, "WORKFLOWY_AUTH_FAILED", "WorkFlowy API klíč je neplatný")
        if response.status_code == 429:
            raise AppError(429, "WORKFLOWY_RATE_LIMIT", "WorkFlowy API rate limit — zkus to za minutu nebo použij cache")
        if not response.is_success:
            raise AppError(502, "WORKFLOWY_ERROR", f"WorkFlowy API chyba: {response.status_code}")

        return response.json()

    async def get_tree(self, force_refresh: bool = False) -> list[WorkFlowyTreeNode]:
        """Načte celý strom — zkusí export, při rate limitu fallback na rekurzivní /nodes."""
        try:
            flat = await self.get_all_nodes_flat(force_refresh)
            return build_tree(flat)
        except AppError as err:
            if err.code == "WORKFLOWY_RATE_LIMIT":
                return await self._get_tree_recursive("None")
            raise

    async def get_all_nodes_flat(self, force_refresh: bool = False) -> list[WorkFlowyExportNode]:
        """Načte ploché pole všech uzlů přes /nodes-export (rate limit: 1/min, proto cache)."""
        if not force_refresh and self._cache and (time.time() - self._cache_timestamp) < self._cache_ttl:
            return self._cache["nodes"]

        response = await self._request("GET", "/nodes-export")
        nodes = [_parse_export_node(n) for n in response["nodes"]]
        self._cache = {"nodes": nodes}
        self._cache_timestamp = time.time()
        return nodes

    async def _get_tree_recursive(self, parent_id: str, depth: int = 0) -> list[WorkFlowyTreeNode]:
        """Rekurzivní načítání stromu přes /nodes (bez rate limitu, ale pomalejší)."""
        if depth > 5:
            return []

        children = await self.get_children(parent_id)
        result: list[WorkFlowyTreeNode] = []

        for child in children:
            grandchildren = await self._get_tree_recursive(child.id, depth + 1)
            result.append(WorkFlowyTreeNode(
                id=child.id, name=child.name, note=child.note, priority=child.priority,
                created_at=child.created_at, modified_at=child.modified_at,
                completed_at=child.completed_at, data=child.data, children=grandchildren,
            ))

        return sorted(result, key=lambda n: n.priority)

    async def get_node(self, node_id: str) -> WorkFlowyNode:
        """Načte konkrétní uzel."""
        response = await self._request("GET", f"/nodes/{node_id}")
        return _parse_node(response["node"])

    async def get_children(self, parent_id: str) -> list[WorkFlowyNode]:
        """Načte děti uzlu."""
        response = await self._request("GET", f"/nodes?parent_id={parent_id}")
        return [_parse_node(n) for n in response["nodes"]]

    def invalidate_cache(self) -> None:
        self._cache = None


def _parse_node(data: dict[str, Any]) -> WorkFlowyNode:
    return WorkFlowyNode(
        id=data["id"], name=data["name"], note=data.get("note"),
        priority=data.get("priority", 0), created_at=data.get("createdAt", 0),
        modified_at=data.get("modifiedAt", 0), completed_at=data.get("completedAt"),
        data=data.get("data"),
    )


def _parse_export_node(data: dict[str, Any]) -> WorkFlowyExportNode:
    return WorkFlowyExportNode(
        id=data["id"], name=data["name"], note=data.get("note"),
        priority=data.get("priority", 0), created_at=data.get("createdAt", 0),
        modified_at=data.get("modifiedAt", 0), completed_at=data.get("completedAt"),
        data=data.get("data"), parent_id=data.get("parent_id"),
    )


def build_tree(nodes: list[WorkFlowyExportNode]) -> list[WorkFlowyTreeNode]:
    """Rekonstruuje strom z plochého pole."""
    node_map: dict[str, WorkFlowyTreeNode] = {}
    roots: list[WorkFlowyTreeNode] = []

    for node in nodes:
        node_map[node.id] = WorkFlowyTreeNode(
            id=node.id, name=node.name, note=node.note, priority=node.priority,
            created_at=node.created_at, modified_at=node.modified_at,
            completed_at=node.completed_at, data=node.data, children=[],
        )

    for node in nodes:
        tree_node = node_map[node.id]
        if node.parent_id and node.parent_id in node_map:
            node_map[node.parent_id].children.append(tree_node)
        else:
            roots.append(tree_node)

    for tree_node in node_map.values():
        tree_node.children.sort(key=lambda n: n.priority)

    return sorted(roots, key=lambda n: n.priority)
